File: data-transformer/src/models/condition.ts
import { Exclude, Expose, Transform } from "class-transformer";
import type { Publication } from "./secondary/publication";
import { assignHeaderIds, cleanupHTML, stepHeaders } from "./utils/html";
import { mapToRule } from "./secondary/rule";

export class Condition {
    @Expose({ name: "_id" }) fvttId!: string;
    name!: string;
    type!: string;

    @Transform(({ obj }) => obj.system.group)
    @Expose()
    group?: string;

    @Transform(({ obj }) => {
        let description = obj.system.description.value;
        description = stepHeaders(description);
        description = assignHeaderIds(description);
        description = cleanupHTML(description);
        return description;
    })
    @Expose()
    description!: string;

    @Transform(({ obj }) => {
        type ConditionDuration = {
            expiry?: boolean;
            perpetual?: boolean;
            text?: string;
            unit: string;
            value: number;
        };
        const durationObj = obj.system.duration as ConditionDuration;
        return durationObj;
    })
    @Expose()
    duration!: any;

    @Transform(({ obj }) => obj.system.overrides)
    @Expose()
    overrides?: string[];

    @Transform(({ obj }) => obj.system.publication)
    @Expose()
    publication!: Publication;

    @Transform(({ obj }) => obj.system.removable)
    @Expose()
    removable?: boolean;

    @Transform(({ obj }) => obj.system.rules.map((v: any) => mapToRule(v)))
    @Expose()
    rules!: any[];

    @Transform(({ obj }) => obj.system.slug)
    @Expose()
    slug!: string;

    @Transform(({ obj }) => obj.system.value.value)
    @Expose()
    value?: number;

    @Exclude() effects!: null;
    @Exclude() _stats!: null;
    @Exclude() img!: null;
    @Exclude() system!: null;
}

export class ConditionDuration {
    perpetual?: boolean;
    text?: string;
    unit!: string;
    value!: number;
}

File: backend/src/models/primary/condition.rs
use crate::helpers::null_to_default;
use crate::models::{LinkPreview, Publication, Rule, Summary, SummaryData};
use crate::storeable::Storeable;
use bson::oid::ObjectId;
use serde::{Deserialize, Serialize};
use specta::Type;

#[derive(Serialize, Deserialize, Debug, Clone, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Condition {
    #[serde(rename = "_id")]
    pub mongo_id: ObjectId,
    pub fvtt_id: String,
    pub name: String,

    pub group: Option<String>,
    pub description: String,
    pub duration: ConditionDuration,
    #[serde(default, deserialize_with = "null_to_default")]
    pub overrides: Vec<String>,
    pub publication: Publication,
    pub removable: Option<bool>,
    pub rules: Vec<Rule>,
    pub slug: String,
    pub value: Option<u8>,
    #[serde(rename = "type")]
    pub data_type: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ConditionDuration {
    pub expiry: Option<bool>,
    pub perpetual: Option<bool>,
    pub text: Option<String>,
    pub unit: String,
    pub value: i8,
}

impl Storeable for Condition {
    fn table_name() -> &'static str {
        "condition"
    }
}

impl From<Condition> for Summary {
    fn from(value: Condition) -> Self {
        let data = vec![SummaryData::String {
            value: value.name.clone(),
            link: None,
            tooltip: None,
            abbreviation: None,
        }];

        Self {
            id: value.fvtt_id,
            name: value.name,
            slug: value.slug,
            data,
        }
    }
}

impl From<Condition> for LinkPreview {
    fn from(value: Condition) -> Self {
        Self {
            slug: value.slug,
            name: value.name,
            description: value.description,
            rarity: None,
            traits: Vec::new(),
        }
    }
}
